x-pack/solutions/security/plugins/elastic_assistant/server/routes/agent_builder_execute.ts

```diff
@@ -0,0 +1,744 @@
+  );
+};
+
+// Helper function to check if result is alert counts
+const isAlertCountsResult = (step: { tool_id?: string }, result: { data?: unknown }): boolean => {
+  return Boolean(
+    'tool_id' in step &&
+      step.tool_id === 'core.security.alert_counts' &&
+      result.data &&
+      typeof result.data === 'object' &&
+      'result' in result.data
+  );
+};
+
+// Helper function to check if result is open and acknowledged alerts
+const isOpenAndAcknowledgedAlertsResult = (
+  step: { tool_id?: string },
+  result: { data?: unknown }
+): boolean => {
+  return Boolean(
+    'tool_id' in step &&
+      step.tool_id === 'core.security.open_and_acknowledged_alerts' &&
+      result.data &&
+      typeof result.data === 'object' &&
+      'alerts' in result.data &&
+      Array.isArray(result.data.alerts)
+  );
+};
+
+// Helper function to handle streaming execution
+const executeStreaming = async ({
+  onechatServices,
+  connectorId,
+  conversationId,
+  nextInput,
+  request,
+  abortSignal,
+  contentReferencesStore,
+  onNewReplacements,
+  onLlmResponse,
+  telemetry,
+  actionTypeId,
+  startTime,
+  logger,
+  response,
+  isEnabledKnowledgeBase,
+}: {
+  onechatServices: OnechatPluginStart;
+  connectorId: string;
+  conversationId?: string;
+  nextInput: RoundInput;
+  request: KibanaRequest<unknown, unknown, ExecuteConnectorRequestBody>;
+  abortSignal: AbortSignal;
+  contentReferencesStore: ContentReferencesStore;
+  onNewReplacements: (newReplacements: Replacements) => void;
+  onLlmResponse?: (
+    content: string,
+    traceData?: Message['traceData'],
+    isError?: boolean
+  ) => Promise<void>;
+  telemetry: AnalyticsServiceSetup;
+  actionTypeId: string;
+  startTime: number;
+  logger: Logger;
+  response: KibanaResponseFactory;
+  isEnabledKnowledgeBase: boolean;
+}): Promise<StreamResponseWithHeaders> => {
+  logger.debug(`🚀 [AGENT_BUILDER] Starting streaming chat execution`);
+
+  // Initialize stream factory and get response headers
+  const {
+    end: streamEnd,
+    push,
+    responseWithHeaders,
+  } = streamFactory<{ type: string; payload: string }>(request.headers, logger, false, false);
+
+  let didEnd = false;
+  const handleStreamEnd = (finalResponse: string, isError = false) => {
+    if (didEnd) {
+      return;
+    }
+    if (isError) {
+      telemetry.reportEvent('invoke_assistant_error', {
+        actionTypeId,
+        model: 'unknown',
+        errorMessage: finalResponse,
+        assistantStreamingEnabled: true,
+        isEnabledKnowledgeBase: false,
+        errorLocation: 'handleStreamEnd',
+      });
+    }
+    if (onLlmResponse) {
+      onLlmResponse(finalResponse, {}, isError).catch(() => {});
+    }
+    streamEnd();
+    didEnd = true;
+  };
+
+  // Set up streaming asynchronously (fire and forget)
+  const pushStreamUpdate = async () => {
+    try {
+      // Use the chat service for streaming
+      const chatEvents$ = onechatServices.chat.converse({
```

Comment:
can we do it differently? without using chat converse?
we should avoid duplicating conversations.